### ISSUE TITLE
`-1` in `IndexedArrays` of Awkward v1 (needs a bug-fix release)

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/awkward#2579]
### Version of Awkward Array

1.10.3

### Description and code to reproduce

From Andrew Roberts on Mattermost/LPCPhysDiscuss/Coffea:

it goes invalid after the `ak.where`. All of the arrays that are inputted to the function are valid. here they are in my analysis script (hope the message isn't too long):

a content `[[], [], []]`
a layout
```
<ListOffsetArray64>
    <offsets><Index64 i="[0 0 0 0]" offset="0" length="4" at="0x0000081d54a0"/></offsets>
    <content><EmptyArray/></content>
</ListOffsetArray64>
```
a is_valid `True`
b content `[[None, None, None], [None, None], [None, None, None]]`
b layout
```
<ListOffsetArray64>
    <offsets><Index64 i="[0 3 5 8]" offset="0" length="4" at="0x0000081d8280"/></offsets>
    <content><IndexedOptionArray64>
        <index><Index64 i="[-1 -1 -1 -1 -1 -1 -1 -1]" offset="0" length="8" at="0x000008413640"/></index>
        <content><NumpyArray format="?" shape="0" data="" at="0x000008342eb0"/></content>
    </IndexedOptionArray64></content>
</ListOffsetArray64>
```
b is_valid `True`
c content `[[126, 59.3, 37.9], [54.5, 46.2], [126, 73.5, 32.3]]`
c layout
```
<ListOffsetArray64>
    <offsets><Index64 i="[0 3 5 8]" offset="0" length="4" at="0x0000091e96d0"/></offsets>
    <content><IndexedArray64>
        <index><Index64 i="[5241 5242 5243 55080 55081 85773 85774 85775]" offset="0" length="8" at="0x000009082f70"/></index>
        <content><NumpyArray format="f" shape="90906" data="54.8438 50.4688 20.7812 17.8594 71.375 ... 53.9062 45.125 28.2188 22.6562 17.6094" at="0x00000927dc50">
            <parameters>
                <param key="__doc__">"pt"</param>
            </parameters>
        </NumpyArray></content>
    </IndexedArray64></content>
</ListOffsetArray64>
```
c is_valid `True`
d content `[[False, False, False], [False, False], [False, False, False]]`
d layout
```
<ListArray64>
    <starts><Index64 i="[8 11 13]" offset="0" length="3" at="0x00000901c3d0"/></starts>
    <stops><Index64 i="[11 13 16]" offset="0" length="3" at="0x0000084281e0"/></stops>
    <content><IndexedArray64>
        <index><Index64 i="[-1 -1 -1 -1 -1 -1 -1 -1 5241 5242 5243 55080 55081 85773 85774 85775]" offset="0" length="16" at="0x0000083ca9f0"/></index>
        <content><NumpyArray format="?" shape="90906" data="false false false false false ... false false false false false" at="0x0000094b1790"/></content>
    </IndexedArray64></content>
</ListArray64>
```
d is_valid `False`

thanks for the help!

**Comment by 32robertsa** [comment]
I'll add for the record the relationship between these arrays is `d = ak.where((ak.num(a) > 0), b, ak.full_like(c, False, dtype=bool))`

**Comment by agoose77** [comment]
Here's my interpretation of the v1 code to generate this example:
```python
import awkward as ak
import numpy as np

a = ak.layout.ListOffsetArray64(
    ak.layout.Index64([0, 0, 0, 0]),
    ak.layout.EmptyArray()
)
b = ak.layout.ListOffsetArray64(
    ak.layout.Index64([0, 3, 5, 8]),
    ak.layout.IndexedOptionArray64(
        ak.layout.Index64([-1]*8),
        ak.layout.NumpyArray(np.zeros(0, dtype=np.bool_))
    )
)
c = ak.layout.ListOffsetArray64(
    ak.layout.Index64([0, 3, 5, 8]),
    ak.layout.IndexedArray64(
        ak.layout.Index64([5241, 5242, 5243, 55080, 55081, 85773, 85774, 85775]),
        ak.layout.NumpyArray(np.random.random(90906).astype(np.float32))
    )
)

d = ak.where((ak.num(a) > 0), b, ak.full_like(c, False, dtype=np.bool_))
print(ak.is_valid(d))
print(ak.type(d))
```

I believe I see the problem; the type should be `N * var * ?bool` but instead it's `N * var * bool`, i.e. we're losing the option type!

**Comment by jpivarski** [comment]
It must be an `IndexedArray64` where an `IndexedOptionArray64` is intended (in v1).

If it's happening in `ak.where`, that function was mostly Python, so it might not require a C++ modification.

Here's the v1 `ak.where`:

https://github.com/scikit-hep/awkward/blob/3375b4a7abe3ca38fd74f0e07d0c4a310bcb35c6/src/awkward/operations/structure.py#L1790-L1860

The fact that it goes through `UnionArray8_64.simplify` is suspicious; it could be in the simplification logic. If that's the case, then it would be in the v1 C++.

**Comment by agoose77** [comment]
Closed by #2582
